File: Python/ch04/ReverseLinkedList_Recur.py
# 链表的节点
class Node:
    def __init__(self, data):
        """
        构造方法
        :param data:
        """
        self.data = data
        self.next = None


def reverse(currNode):
    """
    反转
    :param currNode: 当前节点
    :return:
    """
    # 如果当前节点是None或者当前节点的下一个节点是None
    if currNode is None or currNode.next is None:
        # 说明是尾节点
        return currNode

    # 如果前面的都没有命中，说明不是尾节点
    # 下一个节点
    nextNode = currNode.next
    # 下一个节点进入属于他的反转
    newNextNode = reverse(nextNode)
    # 当前节点进行反转
    nextNode.next = currNode
    currNode.next = None

    return newNextNode


def output(headNode):
    """
    输出节点
    :param headNode: 头节点
    :return:
    """
    rnt = []
    while headNode is not None:
        rnt.append(headNode.data)
        headNode = headNode.next
    print(rnt)


if __name__ == '__main__':
    head = Node(0)
    head.next = Node(1)
    head.next.next = Node(2)
    head.next.next.next = Node(3)
    head.next.next.next.next = Node(4)
    head.next.next.next.next.next = Node(5)
    head.next.next.next.next.next.next = Node(6)
    head.next.next.next.next.next.next.next = Node(7)
    head.next.next.next.next.next.next.next.next = Node(8)
    head.next.next.next.next.next.next.next.next.next = Node(9)

    output(head)

    reHead = reverse(head)

    output(reHead)

File: Java/src/ch04/ReverseLinkedList_Recur.java
package ch04;

public class ReverseLinkedList_Recur {

    /**
     * 链表节点
     */
    private static class Node {
        // 节点的值
        int data;
        // next
        Node next;

        /**
         * 构造方法
         * @param data
         */
        Node(int data) {
            this.data = data;
        }
    }

    /**
     * 反转链表
     * @param head head节点
     * @return
     */
    public static Node reverse(Node currNode){
        // 如果当前节点是空，或者当前节点的下一个节点是空，说明已经是尾节点了
        if (currNode == null || currNode.next == null)
            return currNode;
        // 如果前面的没有命中，说明不是尾节点
        // 当前节点的下一个节点
        Node nextNode = currNode.next;
        // 下一个节点进行属于下一个节点反转
        Node newNextNode = reverse(nextNode);
        // 当前节点做属于当前节点的反转操作
        nextNode.next = currNode;
        currNode.next = null;
        // 返回
        return newNextNode;
    }

    /**
     * 输出链表
     * @param head
     */
    public static void print(Node headNode){
        String rnt = "";
        while (headNode!=null) {
            rnt = rnt + headNode.data + " ";
            headNode = headNode.next;
        }
        System.out.println(rnt);
    }

    public static void main(String[] args) {
        Node head = new Node(0);
        head.next = new Node(1);
        head.next.next = new Node(2);
        head.next.next.next = new Node(3);
        head.next.next.next.next = new Node(4);
        head.next.next.next.next.next = new Node(5);
        head.next.next.next.next.next.next = new Node(6);
        head.next.next.next.next.next.next.next = new Node(7);
        head.next.next.next.next.next.next.next.next = new Node(8);
        head.next.next.next.next.next.next.next.next.next = new Node(9);

        print(head);

        Node reHead = reverse(head);

        print(reHead);
    }
}
